processing/src/main/java/io/druid/query/extraction/LookupReferencesManager.java

```diff
@@ -0,0 +1,145 @@
+import com.metamx.common.logger.Logger;
+import io.druid.guice.ManageLifecycle;
+
+import java.io.IOException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * This class provide a basic {@link LookupExtractor} references manager.
+ * It allows basic operations fetching, listing, adding and deleting of {@link LookupExtractor} objects
+ * It is be used by queries to fetch the lookup reference.
+ * It is used by Lookup configuration manager to add/remove or list lookups configuration via HTTP.
+ */
+
+@ManageLifecycle
+public class LookupReferencesManager
+{
+  private static final Logger LOGGER = new Logger(LookupReferencesManager.class);
+  private final ConcurrentMap<String, LookupExtractor> lookupMap = new ConcurrentHashMap();
+  private final Object lock = new Object();
+  private final AtomicBoolean started = new AtomicBoolean(false);
+
+  @LifecycleStart
+  public void start()
+  {
+    synchronized (lock) {
+      if (!started.getAndSet(true)) {
+        LOGGER.info("Started lookup references manager");
+      }
+    }
+  }
+
+  @LifecycleStop
+  public void stop()
+  {
+    synchronized (lock) {
+      if (started.getAndSet(false)) {
+        for (String lookupName : lookupMap.keySet()) {
+          remove(lookupName);
+        }
+      }
+    }
+  }
+
+  /**
+   * @param lookupName namespace of the lookup object
+   * @param lookupExtractor {@link LookupExtractor} implementation reference.
+   *
+   * @return true if the lookup is added otherwise false.
+   *
+   * @throws ISE If the manager is closed.
+   */
+  public boolean addLookup(String lookupName, final LookupExtractor lookupExtractor) throws ISE
+  {
+    assertStarted();
+    return (null == lookupMap.putIfAbsent(lookupName, lookupExtractor));
+  }
+
+  /**
+   * @param lookupName namespace of {@link LookupExtractor} to delete from the reference registry.
+   *                   this function does call the cleaning method {@link LookupExtractor#close()}
+   *
+   * @return true if it is removed
+   */
+  public boolean remove(String lookupName)
+  {
+    LookupExtractor lookupExtractor = lookupMap.remove(lookupName);
+    if (lookupExtractor != null) {
+      try {
+        LOGGER.debug("Removing lookup [%s]", lookupName);
+        lookupExtractor.close();
+      }
+      catch (IOException e) {
+        LOGGER.error(e, "Got exception while closing lookup [%s]", lookupName);
+      }
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * @param lookupName namespace key to fetch the reference of the object {@link LookupExtractor}
+   *
+   * @return reference of {@link LookupExtractor} that correspond the the namespace {@code lookupName}
+   *
+   * @throws ISE if the {@link LookupReferencesManager} is closed or did not start yet
+   */
+  public LookupExtractor get(String lookupName) throws ISE
+  {
+    assertStarted();
+    return lookupMap.get(lookupName);
```

Comment:
does not guarantee that lookupMap contains the referenced lookup (example, shutdown called between assertStarted and this line). Is that OK here?